aidb/vector_database/tasti.py

```diff
@@ -1,21 +1,34 @@
 import numpy as np
 import pandas as pd
-from numba import njit, prange
 from typing import Optional
 
 from aidb.vector_database.vector_database_config import TastiConfig
 
-@njit(parallel=True)
 def get_and_update_dists(x: np.ndarray, embeddings: np.ndarray, min_dists: np.ndarray):
   '''
   :param x: embedding of cluster representatives
   :param embeddings: embeddings of all data
   :min_dists: array to record the minimum distance for each embedding to the embedding of cluster representatives
   '''
-  for i in prange(len(embeddings)):
-    dists = np.sqrt(np.sum((x - embeddings[i]) ** 2))
-    if dists < min_dists[i]:
-      min_dists[i] = dists
+  try:
+    from numba import njit, prange
+
+    @njit(parallel=True)
+    def _get_and_update_dists_numba(x: np.ndarray, embeddings: np.ndarray, min_dists: np.ndarray):
+
+      for i in prange(len(embeddings)):
+        dists = np.sqrt(np.sum((x - embeddings[i]) ** 2))
+        if dists < min_dists[i]:
+          min_dists[i] = dists
+    _get_and_update_dists_numba(x, embeddings, min_dists)
+
+  except:
+    def _get_and_update_dists_no_numba(x: np.ndarray, embeddings: np.ndarray, min_dists: np.ndarray):
+      for i in range(len(embeddings)):
+        dists = np.sqrt(np.sum((x - embeddings[i]) ** 2))
+        if dists < min_dists[i]:
+          min_dists[i] = dists
+    _get_and_update_dists_no_numba(x, embeddings, min_dists)
```

Comment:
I believe you can just define the function and move the shared code out of the try/except block

Reply:
Hi @ddkang , moved shared code outside, let me know if you need any changes.

Reply:
I meant you can simply share the function name and call the function outside of the try/except block

Reply:
You mean like this? - 
```python
def get_and_update_dists(x: np.ndarray, embeddings: np.ndarray, min_dists: np.ndarray):
  '''
  :param x: embedding of cluster representatives
  :param embeddings: embeddings of all data
  :min_dists: array to record the minimum distance for each embedding to the embedding of cluster representatives
  '''
  try:
    from numba import njit, prange

    @njit(parallel=True)
    def _get_and_update_dists(x: np.ndarray, embeddings: np.ndarray, min_dists: np.ndarray):

      for i in prange(len(embeddings)):
        _get_and_update_dists_shared(x, i, min_dists, embeddings)

  except:
    def _get_and_update_dists(x: np.ndarray, embeddings: np.ndarray, min_dists: np.ndarray):
      for i in range(len(embeddings)):
        _get_and_update_dists_shared(x, i, min_dists, embeddings)
  
  _get_and_update_dists(x, embeddings, min_dists)
```

Reply:
yes

Reply:
Okay, pushed the changes.